modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/IgniteSqlCustomAggregationTest.java

```diff
@@ -0,0 +1,255 @@
+import org.apache.ignite.configuration.CacheConfiguration;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.internal.IgniteEx;
+import org.apache.ignite.internal.processors.cache.index.AbstractIndexingCommonTest;
+import org.apache.ignite.internal.processors.query.h2.IgniteH2Indexing;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.h2.api.AggregateFunction;
+import org.junit.Test;
+
+/**
+ * Tests for registration custom aggregation functions.
+ */
+public class IgniteSqlCustomAggregationTest extends AbstractIndexingCommonTest {
+    /** */
+    private static final String CACHE_NAME = "cache";
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        CacheConfiguration<PersonKey, Person> cacheCfg = new CacheConfiguration<PersonKey, Person>(CACHE_NAME)
+            .setIndexedTypes(PersonKey.class, Person.class)
+            .setCacheMode(CacheMode.PARTITIONED)
+            .setBackups(0);
+
+        cfg.setCacheConfiguration(cacheCfg);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids(false);
+    }
+
+    /**
+     * @throws Exception If error.
+     */
+    @Test
+    public void javaAggregateFunctionsCollocated() throws Exception {
+        startGrids(3);
+
+        IgniteEx ignite = startClientGrid(3);
+
+        IgniteCache<PersonKey, Person> cache = ignite.cache(CACHE_NAME);
+
+        loadCache(cache);
+
+        List<List<?>> lists = cache.query(new SqlFieldsQuery(
+            "select companyId,FIRSTVALUE(NAME,AGE), LASTVALUE(NAME,AGE) from \"cache\".Person group by companyId")
+            .setCollocated(true)).getAll();
+
+        assertEquals(10, lists.size());
+
+        for (List<?> row : lists) {
+            Integer companyId = (Integer)row.get(0);
+            String youngest = (String)row.get(1);
+            String oldest = (String)row.get(2);
+
+            assertEquals("name" + companyId, youngest);
+            assertEquals("name9" + companyId, oldest);
+        }
+    }
+
+    /**
+     * @throws Exception If error.
+     */
+    @Test
+    public void javaAggregateFunctionsWithDistinctArgument() throws Exception {
+        startGrids(3);
+
+        IgniteEx ignite = startClientGrid(3);
+
+        for (int i = 0; i < 4; i++) {
+            ((IgniteH2Indexing)grid(i).context().query().getIndexing()).
+                registerAggregateFunction("ACCUMULATE", AccumulateFunction.class);
+        }
+
+        IgniteCache<PersonKey, Person> cache = ignite.cache(CACHE_NAME);
+
+        loadCache(cache);
+
+        List<List<?>> lists = cache.query(new SqlFieldsQuery(
+            "select companyId,ACCUMULATE(DISTINCT companyId) from \"cache\".Person group by companyId")
```

Comment:
please add test to see if `FIRSTVALUE(NAME, DISTINCT AGE)` syntax is supported and what happens to result there.